go-controller/pkg/mock/mock_lsp.go

```diff
@@ -0,0 +1,355 @@
+package mockovn
+
+import (
+	"fmt"
+
+	goovn "github.com/ebay/go-ovn"
+	"github.com/ebay/libovsdb"
+	"k8s.io/klog"
+)
+
+const (
+	LogicalSwitchPortAddresses        string = "LSPAddressesField"
+	LogicalSwitchPortOptions          string = "LSPOptionsField"
+	LogicalSwitchPortDynamicAddresses string = "LSPDynamicAddressesField"
+	LogicalSwitchPortExternalId       string = "LSPExternalIdsField"
+	LogicalSwitchPortPortSecurity     string = "LSPPortSecurityField"
+	FakeUUID                                 = "8a86f6d8-7972-4253-b0bd-ddbef66e9303"
+)
+
+// Get logical switch port by name
+func (mock *MockOVNClient) LSPGet(lsp string) (*goovn.LogicalSwitchPort, error) {
+	var lspCache MockObjectCacheByName
+	var ok bool
+	if lspCache, ok = mock.cache[LogicalSwitchPortType]; !ok {
+		klog.V(5).Infof("cache doesn't have any object of type %s", LogicalSwitchPortType)
+		return nil, goovn.ErrorSchema
+	}
+	var port interface{}
+	if port, ok = lspCache[lsp]; !ok {
+		return nil, goovn.ErrorNotFound
+	}
+	if lspRet, ok := port.(*goovn.LogicalSwitchPort); ok {
+		return lspRet, nil
+	}
+	return nil, fmt.Errorf("Invalid object type assertion for %s", LogicalSwitchPortType)
+}
```

Comment:
This is all vastly more complicated than I would have expected a mock interface to be... it is nearly as long as the _actual_ implementation it is mocking. And I have no idea if it's right.

I would have expected something more like:

    func (mock *MockOVNClient) LSPGet(lspName string) (*goovn.LogicalSwitchPort, error) {
            lsp, exists := mock.logicalSwitchPorts[lspName]
            if !exists {
                    return nil, goovn.ErrorNotFound
            }
            return lsp, nil
    }

If go-ovn is not designed to be easily mocked (which seems to be part of the problem), then you can fix that in `OVNInterface` so that, eg, you don't have to actually use `*goovn.OvnCommand` in your mocks. eg: 

```
type OVNInterface interface {
        ...
        LSPAdd(ls string, lsp string) (OVNInterfaceOVNCommand, error)
        ...
}
```

and then have

```
func (ovnclient *OVNClient) LSPAdd(ls string, lsp string) (OVNInterfaceOVNCommand, error) {
        return ovnclient.client.LSPAdd(ls, lsp)
}

func (mock *MockOVNClient) LSPAdd(lsName string, lspName string) (OVNInterfaceOVNCommand, error) {
        return MockOVNCommand(func() error {
                ls, exists := mock.logicalSwitches[lsName]
                if !exists {
                        return goovn.ErrorNotFound
                }
                lsp, exists := mock.logicalSwitchPorts[lspName]
                if !exists {
                        return goovn.ErrorNotFound
                }
                ls.ports = append(ls.ports, lsp)
                return nil
        }), nil
}
```

Reply:
So the added complexity is to keep the implementation generic for all objects. Once we get to the lsp cache, it's exactly what you have in your version of `LSPGet`. In some ways, this is to enable future mocks of other things like address sets, lsrs etc. Making the `goovn.OvnCommand` an Interface is a good idea, I just don't see if we need to go to the lengths of mocking the hierarchical structure necessarily. i.e. having to have a ls first before having a lsp. Because then we need to create a lot more mock objects in the test logic.

Reply:
> I just don't see if we need to go to the lengths of mocking the hierarchical structure necessarily. i.e. having to have a ls first before having a lsp.

That was just for the example of mocking `LSPAdd`; you can't add a logical switch port unless you have a logical switch to add it to, and any test case that ends up calling `LSPAdd` is likely to make other calls that operate on the logical swich directly as well.

Reply:
The way the mock lsp works right now, it doesn't really validate whether the switch exists or not. For purposes of unit-testing, I think that is good enough, because that means we don't necessarily need to set up the entire control plane via mocks. Making `goovn.Command` an Interface does get rid of the code that un-necessarily creates the fields of `goovn.Command` in the return values. I do like that.

Reply:
@abhat You may want to consider using https://github.com/vektra/mockery to generate mocks for the interfaces. I can certainly walk through what I have done for some of my Unit Test PRs